drivers/i2c/i2c_nrfx_twi.c

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (c) 2017 - 2018, Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
+#include <i2c.h>
+#include <nrfx_twi.h>
+
+#define SYS_LOG_DOMAIN "i2c_nrfx_twi"
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_I2C_LEVEL
+#include <logging/sys_log.h>
+
+#if !(CONFIG_I2C_NRFX_PINS_OVERRIDE_DTS)
+#define CONFIG_I2C_0_NRF_SDA_PIN CONFIG_I2C_0_DTS_NRF_SDA_PIN
+#define CONFIG_I2C_0_NRF_SCL_PIN CONFIG_I2C_0_DTS_NRF_SCL_PIN
+#define CONFIG_I2C_1_NRF_SDA_PIN CONFIG_I2C_1_DTS_NRF_SDA_PIN
+#define CONFIG_I2C_1_NRF_SCL_PIN CONFIG_I2C_1_DTS_NRF_SCL_PIN
+#endif
+
+struct i2c_nrfx_twi_data {
+	struct k_sem sync;
+	volatile nrfx_err_t res;
+};
+
+struct i2c_nrfx_twi_config {
+	nrfx_twi_t twi;
+};
+
+static inline struct i2c_nrfx_twi_data *get_dev_data(struct device *dev)
+{
+	return dev->driver_data;
+}
+
+static inline
+const struct i2c_nrfx_twi_config *get_dev_config(struct device *dev)
+{
+	return dev->config->config_info;
+}
+
+static int i2c_nrfx_twi_transfer(struct device *dev, struct i2c_msg *msgs,
+				 u8_t num_msgs, u16_t addr)
+{
+	for (size_t i = 0; i < num_msgs; i++) {
+		nrfx_twi_xfer_desc_t cur_xfer = {
+			.p_primary_buf  = msgs[i].buf,
+			.primary_length = msgs[i].len,
+			.address	= addr,
+			.type		= (msgs[i].flags & I2C_MSG_READ) ?
+					  NRFX_TWI_XFER_RX : NRFX_TWI_XFER_TX
+		};
+
+		nrfx_err_t res = nrfx_twi_xfer(&get_dev_config(dev)->twi,
+					       &cur_xfer,
+					       (msgs[i].flags & I2C_MSG_STOP) ?
+					       0 : NRFX_TWI_FLAG_TX_NO_STOP);
+		if (res != NRFX_SUCCESS) {
+			return -EIO;
+		}
+
+		k_sem_take(&(get_dev_data(dev)->sync), K_FOREVER);
+		if (get_dev_data(dev)->res != NRFX_SUCCESS) {
+			return -EIO;
```

Comment:
What about adding something like
`SYS_LOG_ERR("Error %d occurred for message %d", get_dev_data(dev)->res, i);`
one line above?

`res` could be copied to a local variable to avoid re-read (this field is volatile).

Reply:
Done